src/components/tl/ucp/gather/gather_knomial.c

```diff
@@ -9,135 +9,195 @@
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                        task, out);
+                    } else {
+                        /*
+                         in this case, data must be split in two
+                         at the destination buffer
+                        */
+                        msg_size = data_size * (tsize - peer);
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                      task, out);
+                        msg_size = data_size * (num_blocks - (tsize - peer));
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(task->gather_kn.scratch,
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                      task, out);
                     }
                 }
+            }
 
-                if (task->gather_kn.dist == 1) { //check if first passage
-                    msg_size = data_size;
-                    if (rank != root) {
-                        status = ucc_mc_memcpy(task->gather_kn.scratch,
-                                            args->src.info.buffer, msg_size,
-                                            args->src.info.mem_type, mtype);
-                    } else if (!UCC_IS_INPLACE(*args)) {
-                        status = ucc_mc_memcpy(
-                            PTR_OFFSET(task->gather_kn.scratch, data_size * rank),
-                            args->src.info.buffer, msg_size,
-                            args->src.info.mem_type, mtype);
-                    }
+            if ((ct != UCC_COLL_TYPE_REDUCE)  &&
+                ucc_knomial_pattern_loop_first_iteration(p)) {
+                msg_size = data_size;
+                if (rank != root) {
+                    status = ucc_mc_memcpy(task->gather_kn.scratch,
+                                           args->src.info.buffer, msg_size,
+                                           args->src.info.mem_type, mtype);
+                } else if (!UCC_IS_INPLACE(*args)) {
+                    status = ucc_mc_memcpy(
+                        PTR_OFFSET(task->gather_kn.scratch, data_size * rank),
+                        args->src.info.buffer, msg_size,
+                        args->src.info.mem_type, mtype);
+                }
 
-                    if (ucc_unlikely(UCC_OK != status)) {
-                        task->super.status = status;
-                        return;
-                    }
+                if (ucc_unlikely(UCC_OK != status)) {
+                    task->super.status = status;
+                    return;
```

Comment:
just checking: do we need to free ucp task here?

Reply:
no, it's done in finalize